src/ipld/util.rs

```diff
@@ -437,3 +438,259 @@ impl<DB: Blockstore, T: Iterator<Item = Tipset> + Unpin> Stream for ChainStream<
+///
+/// # Arguments
+///
+/// * `db` - A database that implements [`Blockstore`] interface.
+/// * `tipset_iter` - An iterator of [`Tipset`], descending order `$child -> $parent`.
+/// * `stateroot_limit` - An epoch that signifies how far back we need to inspect tipsets, in-depth.
+/// This has to be pre-calculated using this formula: `$cur_epoch - $depth`, where `$depth` is the
+/// number of `[`Tipset`]` that needs inspection.
+#[allow(dead_code)]
+pub fn unordered_stream_chain<
+    DB: Blockstore + Sync + Send + 'static,
+    T: Iterator<Item = Tipset> + Unpin + Send + 'static,
+>(
+    db: Arc<DB>,
+    tipset_iter: T,
+    stateroot_limit: ChainEpoch,
+) -> UnorderedChainStream<DB, T> {
+    let (sender, receiver) = kanal::bounded(2048);
+    let (extract_sender, extract_receiver) = kanal::unbounded();
+    let fail_on_dead_links = true;
+    let seen = Arc::new(Mutex::new(CidHashSet::default()));
+    let handle = UnorderedChainStream::<DB, T>::start_workers(
+        db.clone(),
+        sender.clone(),
+        extract_receiver,
+        seen.clone(),
+        fail_on_dead_links,
+    );
+
+    UnorderedChainStream {
+        seen,
+        db,
+        worker_handle: handle,
+        block_receiver: receiver,
+        queue: Vec::new(),
+        extract_sender,
+        tipset_iter,
+        stateroot_limit,
+        fail_on_dead_links,
+    }
+}
+
+// Stream available graph in unordered search. All reachable nodes are touched and dead-links
+// are ignored.
+pub fn unordered_stream_graph<
+    DB: Blockstore + Sync + Send + 'static,
+    T: Iterator<Item = Tipset> + Unpin + Send + 'static,
+>(
+    db: Arc<DB>,
+    tipset_iter: T,
+    stateroot_limit: ChainEpoch,
+) -> UnorderedChainStream<DB, T> {
+    let (sender, receiver) = kanal::bounded(2048);
+    let (extract_sender, extract_receiver) = kanal::unbounded();
+    let fail_on_dead_links = false;
+    let seen = Arc::new(Mutex::new(CidHashSet::default()));
+    let handle = UnorderedChainStream::<DB, T>::start_workers(
+        db.clone(),
+        sender.clone(),
+        extract_receiver,
+        seen.clone(),
+        fail_on_dead_links,
+    );
+
+    UnorderedChainStream {
+        seen,
+        db,
+        worker_handle: handle,
+        block_receiver: receiver,
+        queue: Vec::new(),
+        tipset_iter,
+        extract_sender,
+        stateroot_limit,
+        fail_on_dead_links,
+    }
+}
+
+impl<DB: Blockstore + Send + Sync + 'static, T: Iterator<Item = Tipset> + Unpin>
+    UnorderedChainStream<DB, T>
+{
+    fn start_workers(
+        db: Arc<DB>,
+        block_sender: Sender<anyhow::Result<Block>>,
+        extract_receiver: Receiver<Cid>,
+        seen: Arc<Mutex<CidHashSet>>,
+        fail_on_dead_links: bool,
+    ) -> JoinHandle<anyhow::Result<()>> {
+        task::spawn(async move {
+            let mut handles = JoinSet::new();
+
+            for _ in 0..num_cpus::get() {
```

Comment:
Should I observe some speedup/slowdown if I increase/decrease the number of cores here? For example testing on a `calibnet` snapshot, I don't see much performance. Or should I run the benchmark command on a `mainnet` snapshot maybe?

Reply:
Calibnet is too small of a snapshot, yeah. I have tested this and this setup worked better than others.